### PR TITLE
Populate contract and entity history

### LINKED ISSUE #2687: Populate entity history
### Problem

During review of PR https://github.com/hashgraph/hedera-mirror-node/pull/2666, an edge case was discovered that caused the design to be rethought. The scenario occurs when multiple updates occur for the same entity within the same record file, excluding dummy `EntityId` updates. As a result, the decision was made to move the entity history population from a trigger to application logic and to defer this change to a follow up item.

### Solution

See https://github.com/hashgraph/hedera-mirror-node/pull/2666#discussion_r728240261

### Alternatives

See https://github.com/hashgraph/hedera-mirror-node/pull/2666#discussion_r728208102

### PR DESCRIPTION
**Pull request by steven-sheehy**
**Description**:
- Add a generic way to create upsert SQL by dynamically generating it from table metadata
- Add support for populating contract and entity history
- Create an index on temporary tables to improve upsert performance
- Increase PostgreSQL temporary buffers to 256MB to improve COPY performance

**Related issue(s)**:

Fixes #2687  

**Notes for reviewer**:

**Checklist**

- [x] Documented (Code comments, README, etc.)
- [x] Tested (unit, integration, etc.)

### CODE REVIEW

File: hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/domain/AbstractEntity.java [hunk 1]
```diff
@@ -93,6 +99,10 @@
     @JsonSerialize(using = RangeToStringSerializer.class)
     private Range<Long> timestampRange;
 
+    public void setTimestampRangeUpper(long modifiedTimestamp) {
+        setTimestampRange(Range.closedOpen(getModifiedTimestamp(), modifiedTimestamp));
```
Comment (ijungmann) [inline hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/domain/AbstractEntity.java:103]: Not a big fan of having `getModifiedTimestamp()` next to `modifiedTimestamp` with different values.  Could we call this variable something like `timestampRangeUpper` or `endTimestamp`?

File: hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/repository/upsert/UpsertQueryGeneratorFactory.java [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+        String className = domainClass.getSimpleName() + UpsertQueryGenerator.class.getSimpleName();
+        return existingGenerators.stream()
+                .filter(u -> u.getClass().getSimpleName().equals(className))
+                .findFirst()
+                .orElseGet(() -> create(domainClass));
+    }
+
+    private UpsertQueryGenerator create(Class<?> domainClass) {
+        UpsertEntity upsertEntity = createEntity(domainClass);
+        log.info("{}", upsertEntity);
```
Comment (Nana-EC) [inline hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/repository/upsert/UpsertQueryGeneratorFactory.java:78]: nit:
```suggestion
        log.info("Creating {}", upsertEntity);
```

**Review by ijungmann — COMMENTED** [review]
Looks really good, going to give it a second pass, only noticed one nit.

File: hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/domain/AbstractEntity.java (hunk 1 above)
Comment (steven-sheehy) [reply hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/domain/AbstractEntity.java]: Renamed

File: hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/parser/batch/BatchUpserter.java [hunk 3]
```diff
@@ -43,8 +43,11 @@
 @Log4j2
 public class BatchUpserter extends BatchInserter {
 
+    private static final String TEMP_BUFFERS_SQL = "set temp_buffers = '256MB'";
```
[the remark this answers is not in the document]
Comment (steven-sheehy) [reply hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/parser/batch/BatchUpserter.java]: Made it a property

File: hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/parser/batch/CompositeBatchPersister.java [hunk 4]
```diff
@@ -71,26 +54,27 @@ public void persist(Collection<? extends Object> items) {
...
-        String className = domainClass.getSimpleName() + UpsertQueryGenerator.class.getSimpleName();
-        return upsertQueryGenerators.stream()
-                .filter(u -> u.getClass().getSimpleName().equals(className))
-                .findFirst()
-                .orElseThrow();
+    private BatchPersister create(Class<?> domainClass) {
+        Entity entity = AnnotationUtils.findAnnotation(domainClass, Entity.class);
+        Upsertable upsertable = AnnotationUtils.findAnnotation(domainClass, Upsertable.class);
+
+        if (entity == null) {
```
[the remark this answers is not in the document]
Comment (steven-sheehy) [reply hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/parser/batch/CompositeBatchPersister.java:68]: Not a hot code path as it's loaded once on startup, but made the adjustment.

File: hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/repository/upsert/UpsertQueryGeneratorFactory.java (hunk 2 above)
Comment (steven-sheehy) [reply hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/repository/upsert/UpsertQueryGeneratorFactory.java]: Changed

**Review by ijungmann — APPROVED** [review]
LGTM

**Review by Nana-EC — APPROVED** [review]
LGTM

**Review by xin-hedera — APPROVED** [review]
LGTM

**Pull request merged**